### PR TITLE
feat(context): add more utils to resolve valueOrPromises

### PR DESCRIPTION
**Pull request by raymondfeng**
This PR is spin-off from https://github.com/strongloop/loopback-next/pull/983. It extracts a few utilities to help handle ValueOrPromise.

## Checklist

- [x] `npm test` passes on your machine
- [x] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- [x] API Documentation in code was updated
- [ ] Documentation in [/docs/site](../tree/master/docs/site) was updated
- [ ] Affected artifact templates in `packages/cli` were updated
- [ ] Affected example projects in `examples/*` were updated

### CODE REVIEW

File: packages/context/src/resolver.ts [hunk 1]
```diff
@@ -56,51 +56,20 @@ export function instantiateClass<T>(
...
-        debug('Injected properties for %s:', ctor.name, propertiesOrPromise);
-      }
-      // Inject the properties asynchronously
-      return inst.then(obj => Object.assign(obj, propertiesOrPromise));
-    } else {
-      // Inject the properties synchronously
-      return Object.assign(inst, propertiesOrPromise);
+    return new ctor(...args);
+  });
+  return resolveValueOrPromise(propertiesOrPromise, props => {
```
Comment (shimks) [inline packages/context/src/resolver.ts]: is there a reason as to why we don't use a promise style here?

Comment (raymondfeng) [reply packages/context/src/resolver.ts]: We need to do the work synchronously or asynchronously based on the return value. The usage of `ValueOrPromise` to allow us to call `context.getSync()` and `context.get()`.

File: packages/context/src/value-promise.ts [hunk 2]
```diff
@@ -218,3 +217,42 @@ export function tryWithFinally<T>(
...
+  resolver: (sourceVal: T) => ValueOrPromise<V | undefined>,
+  evaluator: (sourceVal: T, targetVal: V | undefined) => boolean,
+): ValueOrPromise<V | undefined> {
+  const next = source.next();
+  if (next.done) return undefined; // End of the iterator
+  const sourceVal = next.value;
+  const valueOrPromise = resolver(sourceVal);
+  return resolveValueOrPromise(valueOrPromise, v => {
+    if (evaluator(sourceVal, v)) return v;
+    else return resolveUntil(source, resolver, evaluator);
```
Comment (bajtos) [inline packages/context/src/value-promise.ts]: I believe this will cause stack overflow error when all values are resolved synchronously and the Iterator returns enough items to overflow the stack.

I am not sure if there is any elegant way how to keep supporting both sync/async and using `resolveValueOrPromise` under the hood. I see two ways forward that should be relatively straightforward.
 - Modify `resolveUntil` to always work asynchronously - ensure the recursive call of `resolveUntil` is always deferred to the next tick of event loop or promise micro-queue. This will allow us to use `resolveValueOrPromise`.
 - Rework the code to iterate inside the same function, drop `resolveValueOrPromise`. This will preserve sync/async semantics.

Before you start working on this, please add a test that's triggering stack overflow error - this will verify the assumptions and serve as a proof that the new version works correctly.

File: packages/context/src/value-promise.ts [hunk 3]
```diff
@@ -218,3 +217,42 @@ export function tryWithFinally<T>(
...
 }
+
+/**
+ * Resolve an iterator of source values into a result until the evaluator
+ * returns `true`
+ * @param source The iterator of source values
+ * @param resolver The resolve function that maps the source value to a result
+ * @param evaluator The evaluate function that decides when to stop
+ */
+export function resolveUntil<T, V>(
```
Comment (bajtos) [inline packages/context/src/value-promise.ts]: I find the name `resolveUntil` confusing. In my mind, we _resolve_ values from a context (IoC container), but there is no context involved here. I am proposed to rename this method to `mapUntil`.

File: packages/context/src/value-promise.ts [hunk 4]
```diff
@@ -218,3 +217,42 @@ export function tryWithFinally<T>(
...
+  });
+}
+
+/**
+ * Resolve a value or promise with a function that produces a new value or
+ * promise
+ * @param valueOrPromise The value or promise
+ * @param resolver A function that maps the source value to a value or promise
+ */
+export function resolveValueOrPromise<T, V>(
```
Comment (bajtos) [inline packages/context/src/value-promise.ts]: I find the name `resolveValueOrPromise` confusing. In my mind, we _resolve_ values from a context (IoC container), but there is no context involved here. I am proposed to rename this function to `mapValueOrPromise` or perhaps `transformValueOrPromise`.

OTOH, it is true that Promises can be resolved too, so maybe "resolve" is no that bad.

@strongloop/sq-lb-apex Thoughts?

File: packages/context/test/unit/value-promise.unit.ts [hunk 5]
```diff
@@ -205,3 +212,78 @@ describe('resolveMap', () => {
...
+  });
+
+  it('resolves a promise', async () => {
+    const result = await resolveValueOrPromise('a', v =>
+      Promise.resolve(v && v.toUpperCase()),
+    );
+    expect(result).to.eql('A');
+  });
+
+  it('handles a rejected promise', () => {
```
Comment (bajtos) [inline packages/context/test/unit/value-promise.unit.ts]: Please add another test to verify how sync errors are handled.

```ts
resolveValueOrPromise('a', v => {throw new Error(v);});
```

File: packages/context/src/value-promise.ts (hunk 3 above)
Comment (raymondfeng) [reply packages/context/src/value-promise.ts]: I choose `resolveUntil` to be consistent with other methods such as `resolveList`.

File: packages/context/src/value-promise.ts (hunk 4 above)
Comment (raymondfeng) [reply packages/context/src/value-promise.ts]: I renamed it to `transformValueOrPromise`

File: packages/context/test/unit/value-promise.unit.ts (hunk 5 above)
Comment (raymondfeng) [reply packages/context/test/unit/value-promise.unit.ts]: Added

**Comment by raymondfeng** [comment]
> I believe this will cause stack overflow error when all values are resolved synchronously and the Iterator returns enough items to overflow the stack.

Good catch. I rewrote the implementation to avoid that. A few tests are added to cover the corner case.

**Comment by raymondfeng** [comment]
@bajtos PTAL

File: packages/context/src/value-promise.ts (hunk 3 above)
Comment (bajtos) [reply packages/context/src/value-promise.ts]: Fair enough 👍

File: packages/context/src/value-promise.ts [hunk 6]
```diff
@@ -218,3 +217,55 @@ export function tryWithFinally<T>(
...
+): ValueOrPromise<V | undefined> {
+  // Do iteration in loop for synchronous values to avoid stack overflow
+  while (true) {
+    const next = source.next();
+    if (next.done) return undefined; // End of the iterator
+    const sourceVal = next.value;
+    const valueOrPromise = resolver(sourceVal);
+    if (isPromiseLike(valueOrPromise)) {
+      return valueOrPromise.then(v => {
+        if (evaluator(sourceVal, v)) {
```
Comment (bajtos) [inline packages/context/src/value-promise.ts]: Should we allow `evaluator` to return a `Promise<boolean>` too?

Comment (raymondfeng) [reply packages/context/src/value-promise.ts]: We can add it if a use case comes up.
